metrics/metrics/src/main/java/io/helidon/metrics/Registry.java

```diff
@@ -82,763 +51,82 @@ public static Registry create(Type type) {
-        if (candidateType != MetricType.INVALID) {
-            return candidateType;
-        }
-         /*
-         * A metric could be passed as a lambda, in which case neither
-         * MetricType.from() nor, for example, Counter.class.isAssignableFrom,
-         * works. Check each specific metric class using instanceof.
-         */
-        return Stream.of(Counter.class, Gauge.class, Histogram.class, Meter.class, Timer.class, ConcurrentGauge.class)
-                .filter(clazz -> clazz.isInstance(metric))
-                .map(MetricType::from)
-                .findFirst()
-                .orElse(MetricType.INVALID);
+    protected Registry(Type type) {
+        super(type);
     }
 
-    private static MetricType toType(Metric metric) {
-        // Find subtype of Metric, needed for user-defined metrics
-        Class<?> clazz = metric.getClass();
-        do {
-            Optional<Class<?>> optionalClass = Arrays.stream(clazz.getInterfaces())
-                    .filter(Metric.class::isAssignableFrom)
-                    .findFirst();
-            if (optionalClass.isPresent()) {
-                clazz = optionalClass.get();
-                break;
-            }
-            clazz = clazz.getSuperclass();
-        } while (clazz != null);
-        return MetricType.from(clazz == null ? metric.getClass() : clazz);
+    @Override
+    protected Map<Class<? extends HelidonMetric>, MetricType> prepareMetricToTypeMap() {
+        final Map<Class<? extends HelidonMetric>, MetricType> result = new HashMap<>();
+        result.put(HelidonConcurrentGauge.class, MetricType.CONCURRENT_GAUGE);
+        result.put(HelidonCounter.class, MetricType.COUNTER);
+        result.put(HelidonGauge.class, MetricType.GAUGE);
+        result.put(HelidonHistogram.class, MetricType.HISTOGRAM);
+        result.put(HelidonMeter.class, MetricType.METERED);
+        result.put(HelidonTimer.class, MetricType.TIMER);
+        result.put(HelidonSimpleTimer.class, MetricType.SIMPLE_TIMER);
+        return result;
     }
 
-    private static <T extends HelidonMetric> MetricType toType(Class<T> clazz) {
-        return METRIC_TO_TYPE_MAP.getOrDefault(clazz, MetricType.INVALID);
+    @Override
+    protected Map<MetricType, BiFunction<String, Metadata, HelidonMetric>> prepareMetricFactories() {
+        return Map.of(MetricType.COUNTER, HelidonCounter::create,
+                MetricType.HISTOGRAM, HelidonHistogram::create,
+                MetricType.METERED, HelidonMeter::create,
+                MetricType.TIMER, HelidonTimer::create,
+                MetricType.SIMPLE_TIMER, HelidonSimpleTimer::create,
+                MetricType.CONCURRENT_GAUGE, HelidonConcurrentGauge::create);
     }
 
-    // For testing
-    static Map<Class<? extends HelidonMetric>, MetricType> metricToTypeMap() {
-        return METRIC_TO_TYPE_MAP;
+    // -- declarations which let us keep the methods in the superclass protected; we do not want them public.
```

Comment:
Should they therefore be `final`, since otherwise I can simply subclass _this_ class and make them `public`?

Reply:
Does making them `final` really help?

Yes, it would prevent a user developer from overriding those _exact_ methods with `public` implementations in the subclass, but a developer so determined to expose the information returned from the protected methods on `Registry` could just implement in the subclass some parallel methods which delegate to the protected `Registry` methods, `final` or not.

The goal here was to minimize the declared-as-`public` methods exposed from the new `helidon-metrics-api` module, just to help reduce the _blatantly_ public surface area that might confuse developers _using_ the API `Registry` interface. These methods are used (all of them, I think) internally within the package from `MetricsSupport` or tests in `helidon-metrics`.